SyntheticDataOptions and metrics calculations for NaN replication

Ta7ar opened the pull request:
Added `SyntheticDataOptions` which can be set using `profile_options.set({"synthetic_data.is_enabled": True})`. This enables calculation of metrics specifically needed for `capitalone/synthetic-data`.

Added calculation of metrics needed by `capitalone/synthetic-data` for replicating NaN values in the generated dataset.

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -1739,6 +1750,10 @@ def report(self, report_options=None):
                 report["data_stats"][i]["statistics"]["quantiles"] = quantiles
 
+        if self.options.synthetic_data.is_enabled:
+            report["synthetic_data"] = {
+                "nan_replication": self._nan_replication_metrics
+            }
```
would put this in `report["data_stats"][<column_index>]["nan_replication_metrics"]`

taylorfturner commented on dataprofiler/tests/profilers/profiler_options/test_profiler_options.py:
```diff
@@ -27,7 +27,7 @@ def test_default_profiler_options(self, *mocks):
         for column in profile.options.properties:
             # TODO: remove the check for correlation option once it's updated to True
-            if column == "correlation":
+            if column == "correlation" or column == "synthetic_data":
```
change option name ... `nan_replication_metrics`

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2127,166 @@ def _update_chi2(self):
+        data = pd.DataFrame(clean_samples).apply(pd.to_numeric, errors="coerce")
+
+        for col_id in range(len(self._profile)):
+            null_count = getattr(self._profile[col_id], "null_count")
```
add something like `compiler = self._profile[id]` so you don't have to keep repeating the `self._profile[col_id]` call

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2127,166 @@ def _update_chi2(self):
+
+            # Calculate class priors
+            # i.e probability that a value in target col is or is not NaN
+            sample_size = getattr(self._profile[col_id], "sample_size")
```
see 2164 comment

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2127,166 @@ def _update_chi2(self):
+                    prev_mean_nan, mean_nan, prev_null_count, added_null_count
+                )
+
+            col_name = getattr(self._profile[col_id], "name")
```
see 2164 comment

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2127,166 @@ def _update_chi2(self):
+
+        merged_properties = {}
+        for col_id in range(len(self._profile)):
+            self_null_count = getattr(self._profile[col_id], "null_count")
```
see 2164 comment

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2127,166 @@ def _update_chi2(self):
+                continue
+
+            merged_properties[col_id] = {}
+            self_sample_size = getattr(self._profile[col_id], "sample_size")
```
see 2164 comment

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2127,166 @@ def _update_chi2(self):
+
+            merged_properties[col_id]["class_prior"] = [prior_not_nan, prior_nan]
+
+            col_name = getattr(self._profile[col_id], "name")
```
see 2164 comment

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2127,166 @@ def _update_chi2(self):
+        """
+        counts = {}
+        for col_id in range(len(self._profile)):
+            null_count = getattr(self._profile[col_id], "null_count", 0)
+            sample_size = getattr(self._profile[col_id], "sample_size", 0)
```
see 2164 comment

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2127,166 @@ def _update_chi2(self):
+                merged_properties[col_id]["class_mean"] = other_mean
+
+            elif col_id not in other._nan_replication_metrics:
+                self_mean = self._nan_replication_metrics[col_id]["class_mean"]
```
`self._nan_replication_metrics[col_id]["class_mean"]` repated...

`self._nan_replication_metrics[col_id]["class_mean"]` --> `class_mean = self._nan_replication_metrics[col_id]["class_mean"]` and then just use `class_mean` in the rest of the function

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2127,166 @@ def _update_chi2(self):
+                other_true_count = other_sample_size - other_null_count
+
+                self_mean = self._nan_replication_metrics[col_id]["class_mean"]
+                other_mean = other._nan_replication_metrics[col_id]["class_mean"]
```
see comment on 2270 ... same comment for `other_mean`

taylorfturner commented on dataprofiler/profilers/profiler_options.py:
```diff
@@ -932,6 +932,34 @@ def _validate_helper(self, variable_path="CorrelationOptions"):
         return errors
 
 
+class SyntheticDataOptions(BaseInspectorOptions):
+    """For configuring options for synthetic data generation"""
+
+    def __init__(self):
+        """
+        Initialize options for synthetic data generation
+        :ivar is_enabled: boolean option to enable/disable synthetic data generation options
+        :vartype is_enabled: BooleanOption
+        :ivar replicate_nan: boolean option to enable/disable calculating metrics needed for
+        replicating nan values in synthetic data generator
+        :vartype replicate_nan: BooleanOption
+        """
+
+        self.replicate_nan = BooleanOption(is_enabled=True)
+        super().__init__(is_enabled=False)
+
+    def _validate_helper(self, variable_path="SyntheticDataOptions"):
+        """
+        Validate the options do not conflict and cause errors.
+
+        :param variable_path: current path to variable set.
+        :type variable_path: str
+        :return: list of errors (if raise_error is false)
+        :rtype: list(str)
+        """
+        return super()._validate_helper(variable_path)
+
+
```
as discussed, remove and put `nan_replication_metrics` as `self.nan_replication_metrics` in the  `class StructuredOptions(BaseOption)` in `profiler_options.py`

taylorfturner commented on dataprofiler/profilers/profiler_options.py:
```diff
@@ -1170,6 +1200,7 @@ def __init__(self, null_values=None):
         self.data_labeler = DataLabelerOptions()
         self.correlation = CorrelationOptions()
         self.chi2_homogeneity = BooleanOption(is_enabled=True)
+        self.synthetic_data = SyntheticDataOptions()
```
update to `StructuredOptions` I think

Ta7ar replied:
This options is an attribute of the `StructuredOptions` class (the code highlighted is in the `def __init__` of the `StructuredOptions` class) so it wouldn’t really make sense to use `StructuredOptions` here. Was thinking of a simple `BooleanOptions`.

taylorfturner commented on dataprofiler/profilers/utils.py:
```diff
@@ -759,3 +759,20 @@ def merge_profile_list(list_of_profiles, pool_count=5):
+    :return: Combined mean vector
+    :rtype: np.ndarray
+    """
+    return (mean1 * n1 + mean2 * n2) / (n1 + n2)
```
why not `return (mean1 + mean2) / 2`?

Ta7ar replied:
That equation would only work if `n1 == n2`. If sample sizes the means were calculated from are different then we need to use `(mean1 * n1 + mean2 * n2) / (n1 + n2)`

taylorfturner replied:
Ref: https://www.statisticshowto.com/combined-mean/

taylorfturner replied:
not weighted mean

Ta7ar commented:
Updates in the recent commit include:
* Reduced code redundancy
* Removed `SyntheticDataOptions` class and use `BooleanOptions` class instead
* Renamed options from `synthetic_data` to `null_replication_metrics` (`null` instead of the previous `nan` because `DataProfiler` reports `null_count`, `null_types`, `null_types_index` so `null` sounds more cohesive)
* Change `class_mean` type from `np.ndarray` to `list` (serializable)
* `null_replication_metrics` reported in `data_stats` array
* Updated test cases

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2130,161 @@ def _update_chi2(self):
+        """
+        merged_properties = {}
+        for col_id in range(len(self._profile)):
+            self_compiler = self._profile[col_id]
+            other_compiler = other._profile[col_id]
```
I may have done `self_profile` and `other_profile` but won't hold the PR up over this

Ta7ar commented:
Test coverage:
`dataprofiler/tests/profilers/test_profile_builder.py                                              1612      2    99%`

JGSweets commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2130,161 @@ def _update_chi2(self):
+            partition_not_null = data[~col_is_null]
+
+            # Drop the target column from partitions
+            partition_null = partition_null.drop(columns=col_id)
+            partition_not_null = partition_not_null.drop(columns=col_id)
```
Can we index the values with the boolean without dropping? Calculate the means without dropping?

Ta7ar replied:
The only reason we drop the `col_id` (which is the column where we will be replicating null values in the synthetic data) is because we do not want to have the mean of that column in the mean vector. So we could calculate the mean first and then drop the column but that would be more expensive (calculating the mean of a column we don't need)

JGSweets replied:
I forget though if slicing also copies though.

JGSweets commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2130,161 @@ def _update_chi2(self):
+            prior_not_null = true_count / sample_size
+
+            # Partition data based on whether target column value is null or not
+            col_is_null = data.iloc[:, col_id].isnull()
```
My concern with this call is `isnull` only generates a subset of the nulls determined by the clean operation in struct col. This could cause confusion.

Ta7ar replied:
Could you elaborate on what you mean by "`isnull` only generates a subset of the nulls"? From what I understand the function takes in `clean_sampled_dict` (line 2390) as an argument where each column in the said dict had the `clean_data_and_get_base_stats` (line 470) function called which removes "any nulls in column". So I don't think `isnull` would give a subset, it should give all null values as detected by `clean_data_and_get_base_stats`.

JGSweets commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2130,161 @@ def _update_chi2(self):
+            # Calculate mean vectors of each partition
+            # i.e mean vectors of rows where target col is or is not null
+            mean_not_null, mean_null = (
+                partition_not_null.mean().tolist(),
+                partition_null.mean().tolist(),
```
what if we were to calc this relative to the mean we know for each column against union of null to non-null in the other columns?

JGSweets replied:
Also, do we need to calc the mean of both? We should know the mean of the entire column, and the mean of one partial, hence we could calculate the other without O(n)

JGSweets commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2130,161 @@ def _update_chi2(self):
+                prev_class_means = self._null_replication_metrics[col_id]["class_mean"]
+                prev_mean_not_null, prev_mean_null = prev_class_means
+
+                mean_not_null = utils.combine_means(
```
What if we store sums and counts as oppose to re'calcing? Would that be better? Similar to how we store in numerical stats?

JGSweets reviewed (changes requested):
Would like to validate some ideas prior to approval.

JGSweets commented on dataprofiler/profilers/profiler_options.py:
```diff
@@ -932,6 +932,34 @@ def _validate_helper(self, variable_path="CorrelationOptions"):
         return errors
 
 
+class SyntheticDataOptions(BaseInspectorOptions):
```
The paradigm we've tried to follow previously is for the Options to match the naming of the property or grouping of properties being calculate. In StructuredProfiler it is called `nan_replication_metrics`. I think it would make sense for the options to reflect as: `NanReplicationMetricsOptions`. I'm open to suggestion on this.

I do see a pathway for choosing all synth data options such that it turns off anything that doesn't need to be calculated for synth data.

taylorfturner replied:
I'm open to changing as well. The idea putting it in `StructuredProfiler` is that it is specific to structured data for right now.  I'd rather just create a `SyntheticDataOptions` and not an options class for just this one `nan_replication_metrics`. 

I do see a pathway for `SyntheticDataOptions` as well.

JGSweets commented on dataprofiler/profilers/profiler_options.py:
```diff
@@ -1170,6 +1172,7 @@ def __init__(self, null_values=None):
         self.data_labeler = DataLabelerOptions()
         self.correlation = CorrelationOptions()
         self.chi2_homogeneity = BooleanOption(is_enabled=True)
+        self.null_replication_metrics = BooleanOption(is_enabled=False)
```
nice

Ta7ar commented:
Changes made in latest commit:
* General improvements in runtime and space complexity.
  * `mean_not_null` no longer explicitly calculated by iterating over dataset. Instead `sum_null` is calculated by iterating over rows where target column value is `null`. Then `sum_null` is subtracted from `total_row_sum` (pre calculated) to obtain `sum_not_null` from which `mean_not_null` gets calculated by element wise division.
  * Copies of dataset are no longer created from partitioning data and dropping columns. Instead required calculations take place through explicit dataframe indexing.
* Removed `_get_null_and_sample_counts` and `utils.combine_means` as they are no longer needed.
* Updated test cases to test newly added field `class_sum` in `null_replication_metrics`.
* Indices of null values obtained from `null_types_index` instead of `df[col_id].isnull()`

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2130,171 @@ def _update_chi2(self):
+
+            # Keep only the null indices inside the chunk (reverse index shift)
+            if profile._index_shift is not None:
+                null_indices -= profile._index_shift
+                null_indices = [index for index in null_indices if index >= 0]
```
```suggestion
                null_indices = [index - profile._index_shift for index in null_indices if index >= 0]
```

taylorfturner replied:
or something like 

```
shift = profile._index_shift
[index - shift for index in null_indices if index >= 0]
```

Ta7ar replied:
to condense those two lines into one, the list comprehension would be,
`null_indices = [index - profile._index_shift for index in null_indices if index - profile._index_shift >= 0]`
Could implement it but feels a bit too long for a list comprehension

taylorfturner replied:
oh fair point.. yeah keep as is

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -1438,6 +1438,9 @@ def __init__(self, data, samples_per_update=None, min_true_samples=0, options=No
         self.chi2_matrix = None
 
+        # capitalone/synthetic-data specific metrics
+        self._null_replication_metrics = dict()
```
instantiating class variables with `None` is best practice. Interesting we do it on line 1435. 

https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

taylorfturner replied:
we do it on line 1438 and 1439

taylorfturner commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2130,171 @@ def _update_chi2(self):
+            mean_not_null = mean_not_null.tolist()
+
+            merged_properties[col_id] = {
+                # Array index serves as class label
+                # 0 indicates not null, 1 indicates null
```
nit picky and not going to hold the PR up over it... just move the comments to above the dict and not inside the dict.

JGSweets commented on dataprofiler/profilers/profile_builder.py:
```diff
@@ -2112,6 +2130,174 @@ def _update_chi2(self):
+        :param clean_samples: input cleaned dataset
+        :type clean_samples: dict
+        """
+        data = pd.DataFrame(clean_samples).apply(pd.to_numeric, errors="coerce")
```
is this call still needed? I assume yes for the conversion to numeric? Do we need to apply the entire array? is this expensive?

Ta7ar replied:
Honestly, not too sure about that. Doing it because that is what correlation calculation does and I need the same type of data for null replication as correlation calculation uses.

JGSweets commented on dataprofiler/tests/profilers/test_profile_builder.py:
```diff
@@ -1927,6 +1927,97 @@ def test_unique_row_ratio_empty_profiler(self):
+    def test_null_replication_metrics_calculation(self):
+        data = pd.DataFrame(
+            {
+                "a": [3, 2, np.nan, 7, np.nan],
```
should we validate that the clean, etc. works for nulls that aren't just caught by `np.nan`. e.g. blank spaces, or None?

Ta7ar replied:
As long as the DataProfiler successfully detects null values the calculation shouldn't be affected. Could definitely add that to the tests though.

JGSweets commented on dataprofiler/tests/profilers/test_profile_builder.py:
```diff
@@ -1927,6 +1927,97 @@ def test_unique_row_ratio_empty_profiler(self):
+            }
+        )
+
+        profiler.update_profile(data_2)
```
shouldn't the values after this and the values after `merge` be the same?

Ta7ar replied:
Initially `profiler` has `data`, then update `profiler` with `data_2` (so `profiler` has `data` and `data_2`). Then initialize `profiler2` with `data` then merge it with `profiler`, therefore `merged_profile` contains `2*data+data_2` whereas `profile` after update contains `data+data_2` so it shouldn't be the same.

JGSweets replied:
That makes sense. I thought it was going to be `data + data2`

I might suggest doing that as we should validate that merge / update should provide the same results.

one way would be to do merge prior to update to validate.

JGSweets commented on dataprofiler/tests/profilers/test_profile_builder.py:
```diff
@@ -1927,6 +1927,97 @@ def test_unique_row_ratio_empty_profiler(self):
+        )
+        profile_options = dp.ProfilerOptions()
+        profile_options.set(
+            {
+                "data_labeler.is_enabled": False,
+                "structured_options.multiprocess.is_enabled": False,
+                "null_replication_metrics.is_enabled": True,
+            }
```
nice
